### ISSUE TITLE
Add non-file-based option

### ISSUE THREAD

**Issue by LindyBalboa** [olistrik/goskew#12]
So most printers have start-up, end, pause, filament/toolhead change gcodes. Applying skew correction to these operations is often frowned upon, because they involve the toolhead move close to the end limits. 

Orca Slicer for example brackets object sections of gcode (at least for Bambu printers) in code such as 

```
; start printing object, unique label id: 3169
M624 AQAAAAAAAAA=
...
...
; stop printing object, unique label id: 3169
M625
```

This would lend itself to easy chunking into blocks to limit the scope of the de-skewing. Since this kind of code is very slicer and printer dependent, it would be difficult to cover all the cases within your code. If you add on option for using strings or stdin/out it would be relatively easy to write wrappers around your code for specific flavors of gcode, e.g. breaking the input file up using the above delimiting lines and only running your code. 

Here is an example of some pytohn/pseudocode from chatgpt of what I have in mind

```
import re

# Sample G-code string (normally you would read this from a file)
gcode = """
G1 X0 Y0 Z0
; start printing object, unique label id: 3169
G1 X10 Y10 Z0.3
G1 X20 Y10 Z0.3
; stop printing object, unique label id: 3169
G1 X30 Y10 Z0.3
; start printing object, unique label id: 3170
G1 X10 Y20 Z0.3
G1 X20 Y20 Z0.3
; stop printing object, unique label id: 3170
G1 X30 Y20 Z0.3
"""
XYTAN = 0.1
XZTAN = 0.2
YZTAN = 0.3
skew_arguments = f"--xy={XYTAN} --xz={XZTAN} --yz={YZTAN}"

# Define the regex pattern
pattern = re.compile(r"; start printing object, unique label id: (\d+)\n(.*?)\n; stop printing object, unique label id: \1", re.DOTALL)

# Define a function to process each block of code
def process_block(match):
    unique_id = match.group(1)
    block = match.group(2)
    
    # Process the block (for example, alter the lines)
    processed_block = subprocess(f"goskew {skew_argmuments} {block}")
    
    # Return the modified block with the original comment lines
    return f"; start printing object, unique label id: {unique_id}\n{processed_block}\n; stop printing object, unique label id: {unique_id}"

# Use re.sub with the process_block function to replace matched blocks
processed_gcode = pattern.sub(process_block, gcode)
```

**Comment by olistrik** [comment]
Hi @LindyBalboa! 

I can almost certainly add support for reading and writing via stdio, I'll just need to be sure that I won't break existing usages.

I'm also open to discussing adding custom goSkew instructions as gcode comments, that way rather than writing a wrapper script, you could just add `; @goskew enable` and `; @goskew disable` to disable goskewer for sections that you don't want skewed. Prusa slicer allows customizing the start-up, end, pause, filament/toolhead change gcodes, as well as the before and after layer change gcodes, I'm assuming that other slicers allow this too.

The impact of skewing (or not) all movement commands is something I've considered in the past. Ultimately I decided to skew everything that I could because it was simpler and would produce the most predictable results. I'd also assumed that most printers would have enough extra room on each axis, and the skew correction would be small enough that even if printing to the limits of the machine it still wouldn't hit the end stops. I hadn't considered tool changes though.

If you don't mind sharing, I'm interested in if there's a particular scenario that you're running into issues with? You mentioned Bamboo printers, you might want to look into the M1005 instruction, it seems to do something similar to M852.

Also keep in mind that there have been issues in the past with certain gcode instructions not being skewed (https://github.com/olistrik/goskew/issues/8) and the resulting prints "exploding". Depending on if you're instructions are absolute or relative you may get get drastically different results if you disable skewing for sub sections of gcode.

**Comment by LindyBalboa** [comment]
Hi, the idea with the gcode comments is a good one. I guess there is also the block of code for filament changes and whatnot where that could be disabled at the start and re-enabled afterwards. Bambu and Orca are also Slic3r derivatives and treat this basically like Prusa. 

I didn't have time to check it in detail, but I only tried it on one print so far where Goskew caused my printhead to crash multiple times halfway through the first layer. I checked the gcode and didn't see anything off about it. I can upload it tonight or tomorrow once I am back home. 

As for the M1005 command, it appears to only work on the X1C machines. Customer support people on the forums say "they don't see why it shouldn't work on the P1P/S" but they have no idea. Multiple P users including myself so no effect with that command.

Orca uses absolute by default, I am not sure if relative is even an option.

**Comment by olistrik** [comment]
>  I guess there is also the block of code for filament changes and whatnot where that could be disabled at the start and re-enabled afterwards.

You could likely implement the comments in a few different ways. I think the least amount of work would be to disable gskewer on startup and only enable and disable it at the start and end of each of layer. Then you don't need to worry too much about other operations, assuming that it stops each layer before doing anything else.

If all instructions are absolute then you shouldn't have too many issues with disabling blocks, the only thing I can think of is if it starts / stops extruding outside of the layer block. Even then though, it should only be a matter of where you put the `disable`/`enable` comments.

If you can share both a skewed and unskewed file for reference that would be great! Shame that the M1005 command doesn't work.

**Comment by LindyBalboa** [comment]
So I have printed a few items with skew correction since with no problems. I have an idea that the issue may have been a mismatched profile. The model I had issues with came from MakerWorld as an 3mf with the printer set as an A1. I switched it to P1S and the gcode specified P1S correctly, but maybe I wound up with some messed up settings? I have seen reports of this happening, the results being head crashes. Not sure.

At any rate, on my last calilantern print, all of my errors were below 0.1 degrees, with my XY at 0.002, and I had no issues with startup purges or anything else, so it appears to be working. I still think making enabling/disabling comments available would be a great addition that the community would appreciate. 

I checked and the layers are not interrupted between filament changes. This makes sense, but throws a wrench in your idea to only toggle it on layer start/stop. So I guess the plan would be to enable it at the bottom of the start gcode, and wrap the filament change with a disable/enable pair. I've included a small file that shows a color change within a layer if you care to take a look. 
[Cube_PETG_10m46s.txt](https://github.com/user-attachments/files/16280257/Cube_PETG_10m46s.txt)
